File: design-patterns-exercises/template-method/test/controller/PaymentController.test.ts
import * as request from 'supertest';
import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { PaymentRepository } from '../../src/repository/PaymentRepository';
import { PaymentController } from '../../src/controller/PaymentController';
import {
  CREDIT_BANK_API,
  CREDIT_CARD_API,
  CREDIT_PAYPAL_PAY_API,
  CREDIT_PAYPAL_TOKEN_API,
  PaymentService,
} from '../../src/service/PaymentService';
import { HttpModule, HttpService } from '@nestjs/axios';
import { PaymentRequest } from '../../src/controller/PaymentRequest';
import { mockServer } from '../mock/MockServer';

describe('PaymentController (e2e)', () => {
  let app: INestApplication;
  let paymentRepository: PaymentRepository;
  let httpService: HttpService;

  beforeAll(() => mockServer.listen({ onUnhandledRequest: 'bypass' }));
  afterEach(() => mockServer.resetHandlers());
  afterAll(() => mockServer.close());
  afterAll(() => jest.resetAllMocks());

  beforeEach(async () => {
    paymentRepository = {
      save: jest.fn(),
    };

    const moduleRef: TestingModule = await Test.createTestingModule({
      imports: [HttpModule],
      controllers: [PaymentController],
      providers: [
        PaymentService,
        {
          provide: 'PaymentRepository',
          useValue: paymentRepository,
        },
      ],
    }).compile();

    app = moduleRef.createNestApplication();

    httpService = moduleRef.get<HttpService>(HttpService);

    await app.init();
  });

  it('should pay with creditCard', async () => {
    const httpServiceSpy = jest.spyOn(httpService, 'post');
    const paymentRequest = new PaymentRequest();
    paymentRequest.amount = 100000;
    paymentRequest.method = 'credit card';
    paymentRequest.cardNumber = '378282246310005';
    paymentRequest.cardCvv = '123';
    paymentRequest.cardExpDate = '01/01/2023';

    await request(app.getHttpServer()).post('/payment').send(paymentRequest).expect(200);

    expect(httpServiceSpy).toHaveBeenCalledWith(
      CREDIT_CARD_API,
      expect.objectContaining({
        amount: 100000,
        cardCvv: '123',
        cardExpDate: '01/01/2023',
        cardNumber: '378282246310005',
      }),
    );

    expect(paymentRepository.save).toHaveBeenCalledWith(
      expect.objectContaining({
        amount: 100000,
        paymentMethod: 'credit card',
        providerId: 1,
      }),
    );
  });

  it('should pay with paypal', async () => {
    const httpServiceSpy = jest.spyOn(httpService, 'post');
    const paymentRequest = new PaymentRequest();
    paymentRequest.amount = 100000;
    paymentRequest.method = 'PayPal';
    paymentRequest.paypalEmail = 'jhon@gamil.com';
    paymentRequest.paypalPassword = '123456';

    await request(app.getHttpServer()).post('/payment').send(paymentRequest).expect(200);

    expect(httpServiceSpy).toHaveBeenNthCalledWith(
      1,
      CREDIT_PAYPAL_TOKEN_API,
      expect.objectContaining({ paypalEmail: 'jhon@gamil.com', paypalPassword: '123456' }),
    );

    expect(httpServiceSpy).toHaveBeenNthCalledWith(
      2,
      CREDIT_PAYPAL_PAY_API,
      expect.objectContaining({ amount: 100000, token: 1 }),
    );

    expect(paymentRepository.save).toHaveBeenCalledWith(
      expect.objectContaining({
        amount: 100000,
        paymentMethod: 'PayPal',
        providerId: 1,
      }),
    );
  });

  it('should pay with bank transfer', async () => {
    const httpServiceSpy = jest.spyOn(httpService, 'post');
    const paymentRequest = new PaymentRequest();
    paymentRequest.amount = 100000;
    paymentRequest.method = 'bank transfer';
    paymentRequest.bankAccountNumber = '123456789';
    paymentRequest.bankRoutingNumber = '98765';

    await request(app.getHttpServer()).post('/payment').send(paymentRequest).expect(200);

    expect(httpServiceSpy).toHaveBeenCalledWith(
      CREDIT_BANK_API,
      expect.objectContaining({ amount: 100000, bankAccountNumber: '123456789', bankRoutingNumber: '98765' }),
    );

    expect(paymentRepository.save).toHaveBeenCalledWith(
      expect.objectContaining({
        amount: 100000,
        paymentMethod: 'bank transfer',
        providerId: 1,
      }),
    );
  });

  it('should throw error when invalid method', async () => {
    const paymentRequest = new PaymentRequest();
    paymentRequest.amount = 100000;
    paymentRequest.method = 'Invalid Method';

    await request(app.getHttpServer()).post('/payment').send(paymentRequest).expect(500);

    expect(paymentRepository.save).not.toBeCalled();
  });
});

File: pattern-factory/spring-boot/src/test/java/com/juandaabril/factory/controller/PaymentControllerTest.java
package com.juandaabril.factory.controller;

import static com.juandaabril.factory.service.PaymentService.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.*;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.juandaabril.factory.entity.Payment;
import com.juandaabril.factory.repository.PaymentRepository;
import com.juandaabril.factory.service.PaymentService;
import java.math.BigDecimal;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.test.web.client.MockRestServiceServer;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.web.client.RestTemplate;

@WebMvcTest(PaymentController.class)
@Import(PaymentService.class)
class PaymentControllerTest {
  @Autowired private MockMvc mockMvc;

  @Autowired private ObjectMapper objectMapper;

  @Autowired private RestTemplate restTemplate;

  @MockBean private PaymentRepository paymentRepository;

  private MockRestServiceServer mockServer;

  @Captor private ArgumentCaptor<Payment> paymentCaptor;

  @BeforeEach
  public void setup() {
    mockServer = MockRestServiceServer.createServer(restTemplate);
  }

  @Test
  public void shouldPayWithCreditCard() throws Exception {
    mockServer
        .expect(requestTo(CREDIT_CARD_API))
        .andExpect(method(HttpMethod.POST))
        .andExpect(
            content()
                .string(
                    "{\"amount\":100000,\"cardNumber\":\"378282246310005\",\"cardExpDate\":\"01/01/2023\",\"cardCvv\":\"123\"}"))
        .andRespond(withSuccess("{ \"id\": 1 }", MediaType.APPLICATION_JSON));

    PaymentRequest request = new PaymentRequest();
    request.setAmount(BigDecimal.valueOf(100000));
    request.setMethod("credit card");
    request.setCardNumber("378282246310005");
    request.setCardCvv("123");
    request.setCardExpDate("01/01/2023");

    mockMvc
        .perform(
            post("/payment")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
        .andExpect(status().isOk());

    verify(paymentRepository, times(1)).save(paymentCaptor.capture());

    Payment payment = paymentCaptor.getValue();

    assertThat(payment).isNotNull();
    assertThat(payment.getPaymentMethod()).isEqualTo("credit card");
    assertThat(payment.getProviderId()).isEqualTo(1);
    assertThat(payment.getAmount()).isEqualTo(BigDecimal.valueOf(100000));
  }

  @Test
  public void shouldPayWithPaypal() throws Exception {
    mockServer
        .expect(requestTo(CREDIT_PAYPAL_TOKEN_API))
        .andExpect(method(HttpMethod.POST))
        .andExpect(
            content().string("{\"paypalEmail\":\"jhon@gamil.com\",\"paypalPassword\":\"123456\"}"))
        .andRespond(withSuccess("{ \"id\": 1 }", MediaType.APPLICATION_JSON));

    mockServer
        .expect(requestTo(CREDIT_PAYPAL_PAY_API))
        .andExpect(method(HttpMethod.POST))
        .andExpect(content().string("{\"amount\":100000,\"token\":1}"))
        .andRespond(withSuccess("{ \"id\": 1 }", MediaType.APPLICATION_JSON));

    PaymentRequest request = new PaymentRequest();
    request.setAmount(BigDecimal.valueOf(100000));
    request.setMethod("PayPal");
    request.setPaypalEmail("jhon@gamil.com");
    request.setPaypalPassword("123456");

    mockMvc
        .perform(
            post("/payment")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
        .andExpect(status().isOk());

    verify(paymentRepository, times(1)).save(paymentCaptor.capture());

    Payment payment = paymentCaptor.getValue();

    assertThat(payment).isNotNull();
    assertThat(payment.getPaymentMethod()).isEqualTo("PayPal");
    assertThat(payment.getProviderId()).isEqualTo(1);
    assertThat(payment.getAmount()).isEqualTo(BigDecimal.valueOf(100000));
  }

  @Test
  public void shouldPayWithBankTransfer() throws Exception {
    mockServer
        .expect(requestTo(CREDIT_BANK_API))
        .andExpect(method(HttpMethod.POST))
        .andExpect(
            content()
                .string(
                    "{\"amount\":100000,\"bankAccountNumber\":\"123456789\",\"bankRoutingNumber\":\"98765\"}"))
        .andRespond(withSuccess("{ \"id\": 1 }", MediaType.APPLICATION_JSON));

    PaymentRequest request = new PaymentRequest();
    request.setAmount(BigDecimal.valueOf(100000));
    request.setMethod("bank transfer");
    request.setBankAccountNumber("123456789");
    request.setBankRoutingNumber("98765");

    mockMvc
        .perform(
            post("/payment")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request)))
        .andExpect(status().isOk());

    verify(paymentRepository, times(1)).save(paymentCaptor.capture());

    Payment payment = paymentCaptor.getValue();

    assertThat(payment).isNotNull();
    assertThat(payment.getPaymentMethod()).isEqualTo("bank transfer");
    assertThat(payment.getProviderId()).isEqualTo(1);
    assertThat(payment.getAmount()).isEqualTo(BigDecimal.valueOf(100000));
  }

  @Test
  public void shouldThrowWhenInvalidMethod() throws Exception {
    PaymentRequest request = new PaymentRequest();
    request.setAmount(BigDecimal.valueOf(100000));
    request.setMethod("Invalid Method");

    Assertions.assertThatThrownBy(
            () ->
                mockMvc.perform(
                    post("/payment")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(request))))
        .hasCauseInstanceOf(RuntimeException.class)
        .hasMessageContaining("Método de pago no válido");

    verify(paymentRepository, times(0)).save(any());
  }
}
